notebooks/ch12/linear_case_study.md

```diff
@@ -0,0 +1,1205 @@
+        2.11    2.66   -3.64   -7.32    0.2 ] | cost: 63.70
+    theta: [-204.39    0.93    1.67    0.75   -9.88   -8.66   -6.27    7.49   11.25
+        2.11    2.66   -3.64   -7.32    0.2 ] | cost: 63.70
+    theta: [-204.38    0.93    1.67    0.75   -9.89   -8.66   -6.27    7.49   11.25
+        2.12    2.67   -3.64   -7.31    0.21] | cost: 63.70
+    theta: [-204.38    0.93    1.67    0.75   -9.89   -8.67   -6.27    7.49   11.25
+        2.12    2.67   -3.63   -7.31    0.22] | cost: 63.70
+    theta: [-204.38    0.93    1.67    0.75   -9.89   -8.67   -6.27    7.49   11.26
+        2.13    2.68   -3.62   -7.3     0.23] | cost: 63.70
+    theta: [-204.39    0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23] | cost: 63.70
+    theta: [-204.4     0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23] | cost: 63.70
+    theta: [-204.4     0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23] | cost: 63.70
+    theta: [-204.4     0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23] | cost: 63.70
+    theta: [-204.4     0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23] | cost: 63.70
+    theta: [-204.4     0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23] | cost: 63.70
+    
+
+Looks like gradient descent converged to those theta values! Thus, our linear model is:
+
+$y = -204.4 + 0.93x_1 + ... -7.29x_{13} - 0.64x_{13}$
+
+Let's compare this equation that we obtained to the one we would get if we had used `sklearn`'s LinearRegression model instead.
+
+
+```python
+model = LinearRegression(fit_intercept=False) # We already accounted for it with the bias column
+model.fit(X[:, :14], y)
+print("Coefficients", model.coef_)
+```
+
+    Coefficients [-204.4     0.93    1.67    0.75   -9.88   -8.67   -6.27    7.49   11.26
+        2.13    2.69   -3.62   -7.29    0.23]
+    
+
+The coefficients look pretty similar! Our homemade functions create the same model as an established Python package!
+
+We successfully fit a linear model to our donkey data! Nice!
+
+## Evaluating our Model
+
+Our next step is to evaluate our model's performance on the test set. We need to perform the same data pre-processing steps on the test set as we did on the training set before we can pass it into our model.
+
+
+```python
+donkeys_test = X_test.drop(['Sex', 'WeightAlt'], axis=1)
+donkeys_test = pd.get_dummies(donkeys_test, columns=['BCS', 'Age'])
+donkeys_test = donkeys_test.drop(['BCS_3.0', 'Age_5-10'], axis=1)
+donkeys_test = donkeys_test.assign(bias=1)
+```
+
+
+```python
+# HIDDEN
+donkeys_test = donkeys_test.reindex_axis(['bias'] + list(donkeys_test.columns[:-1]), axis=1)
```

Comment:
prob dont need to hide this. good to remember to add the bias!

Reply:
mmm... the bias is actually in the code block above
that confusing bit of code rearranges the columns so that the bias comes first